File: src/main/resources/static/js/controller/userController.js
'use strict';

app.controller('userController', ['$scope', '$http', '$window', function($scope, $http, $window) {

    $scope.credentials = [];
    $scope.currentUserExists = false;
    $scope.username = "";
    $scope.pass = "";
    $scope.localUserName = "";

    $scope.initPage = function() {
        $scope.credentials = [];
        $scope.currentUserExists = false;
        $scope.localUserName = "";
        $scope.description = "";
        $scope.login = "";
        $scope.password = "";
    }

    /*
     * User
     */
    $scope.createUser = function(username, complexity) {
        $http.post(baseUrl + 'createUser/' + username + "/" + complexity)
            .then(function successCallback(response) {
                if (response.data.userInfo === "USER_ALREADY_EXIST") {
                    window.alert("User " + username + " already exists please create another one")
                    $scope.initPage();
                } else {
                    window.alert("User " + $scope.username + " created successfully")
                    $scope.localUserName = username;
                    $scope.currentUserExists = true;
                    $scope.credentials = [];
                }
            }, function errorCallback(response) {
                console.log("Error while creating user " + username);
                $scope.initPage();
            });
    }

    $scope.loginToServer = function(username, pass) {

        let loginHolder = {
            username: username,
            pass: pass
        };

        let loginRequest = {
            method: 'POST',
            url: baseUrl + 'login',
            headers: {
              'Content-Type': 'application/json'
            },
            data: loginHolder
        };

        $http(loginRequest)
            .then(function successCallback(response) {
                if (response.data.userInfo === "USER_NOT_EXIST") {
                    window.alert("User " + username + " does not exist please create it first");
                    $scope.initPage();
                } else if (response.data.userInfo === "WRONG_PASS") {
                    window.alert("Wrong pass :)");
                    $scope.initPage();
                } else if (response.data.user.credentials.length === 0) {
                    $scope.currentUserExists = true;
                    $scope.localUserName = username;
                    $scope.credentials = [];
                } else {
                    $scope.currentUserExists = true;
                    $scope.localUserName = username;
                    $scope.getCredential(username, pass);
                }
            }, function errorCallback(response) {
                console.log("Error while getting user " + username);
                $scope.initPage();
            });
    }


    /*
     * Credential
     */
    $scope.getCredential = function(username, pass) {
        $http.get(baseUrl + 'getCredential/' + username + "/" + pass)
            .then(function successCallback(response) {
                $scope.credentials = response.data;
                console.log($scope.credentials);
            }, function errorCallback(response) {
                console.log("Error while getting credentials for " + username);
                $scope.initPage();
            });
    }

    $scope.addCredential = function(username) {

        let credential = {
            login: $scope.login,
            password: $scope.password,
            description: $scope.description
        };

        let addCredentialRequest = {
            method: 'POST',
            url: baseUrl + 'addCredential/' + username,
            headers: {
              'Content-Type': 'application/json'
            },
            data: credential
        };

        $http(addCredentialRequest)
            .then(function successCallback(response) {
                console.log("new credential added successfully");
                $scope.credentials = response.data;
            }, function errorCallback(response) {
                console.log("Error while adding new credentials");
                $scope.initPage();
            });
    }

    $scope.deleteCredential = function(credential) {

        let deleteCredentialRequest = {
            method: 'POST',
            url: baseUrl + 'deleteCredential/' + $scope.localUserName,
            headers: {
                'Content-Type': 'application/json'
            },
            data: credential
        };

        $http(deleteCredentialRequest)
            .then(function successCallback(response) {
                console.log("Credential deleted successfully");
                $scope.credentials = response.data;
            }, function errorCallback(response) {
                console.log("Error while deleting credentials");
                $scope.initPage();
            });
    }

    $scope.updateCredential = function(cred, description, login, password) {

        let newCredential = {
            login: cred.login,
            password: cred.password,
            description: cred.description,

        };

        if (description !== undefined) {
            newCredential.description = description;
        }
        if (login !== undefined) {
            newCredential.login = login;
        }
        if (password !== undefined) {
            newCredential.password = password;
        }

        let credentialHolder = {
            oldCredential: cred,
            newCredential: newCredential
        };

        let updateCredentialRequest = {
            method: 'POST',
            url: baseUrl + 'updateCredential/' + $scope.localUserName,
            headers: {
                'Content-Type': 'application/json'
            },
            data: credentialHolder
        };
        $http(updateCredentialRequest)
            .then(function successCallback(response) {
                console.log("Credential update successfully");
                $scope.credentials = response.data;
            }, function errorCallback(response) {
                console.log("Error while updating credentials");
                $scope.initPage();
            });
    }
}]);

File: src/main/java/com/easylife/loki/controller/UserController.java
package com.easylife.loki.controller;

import com.easylife.loki.model.Credential;
import com.easylife.loki.model.User;
import com.easylife.loki.model.UserHolder;
import com.easylife.loki.model.UserInfo;
import com.easylife.loki.service.UserService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.RequestBody;

import java.util.List;


@Controller
public class UserController {

    @Autowired
    UserService userService;

    /**
     * Users
     */
    @PostMapping(value = "/login", consumes = "application/json")
    public ResponseEntity<UserHolder> login(@RequestBody LoginHolder loginHolder) {
        User user = userService.getUser(loginHolder.getUsername());
        if (user == null) {
            return new ResponseEntity<>(new UserHolder(null, UserInfo.USER_NOT_EXIST), HttpStatus.OK);

        } else if (user.getCredentials().size() == 0) {
            return new ResponseEntity<>(new UserHolder(user, UserInfo.SUCCESS), HttpStatus.OK);

        } else if (userService.matchPassword(loginHolder.getUsername(), loginHolder.pass)) {
            return new ResponseEntity<>(new UserHolder(user, UserInfo.SUCCESS), HttpStatus.OK);

        } else {
            return new ResponseEntity<>(new UserHolder(null, UserInfo.WRONG_PASS), HttpStatus.OK);
        }
    }

    @PostMapping(value = "/createUser/{username}/{complexity}")
    public ResponseEntity<UserHolder> createUser(@PathVariable("username") String username, @PathVariable("complexity") Double complexity) {
        if (userService.getUser(username) != null) {
            return new ResponseEntity<>(new UserHolder(null, UserInfo.USER_ALREADY_EXIST), HttpStatus.OK);
        } else {
            userService.createUser(username, complexity);
            return new ResponseEntity<>(new UserHolder(new User(username), UserInfo.SUCCESS), HttpStatus.OK);
        }
    }


    /**
     * Credentials
     */
    @GetMapping(value = "/getCredential/{username}/{pass}")
    public ResponseEntity<Credential[]> getCredential(@PathVariable("username") String username, @PathVariable("pass") String pass) {

        if (userService.matchPassword(username, pass)) {
            return new ResponseEntity<>(this.getCredentials(username), HttpStatus.OK);
        } else {
            return new ResponseEntity<>(null, HttpStatus.OK);
        }
    }

    @PostMapping(value = "/addCredential/{username}", consumes = "application/json")
    public ResponseEntity<Credential[]> addCredential(@PathVariable("username") String username, @RequestBody Credential credential) {
        userService.addCredential(username, credential);
        return new ResponseEntity<>(this.getCredentials(username), HttpStatus.OK);
    }

    @PostMapping(value = "/deleteCredential/{username}", consumes = "application/json")
    public ResponseEntity<Credential[]> deleteCredential(@PathVariable("username") String username, @RequestBody Credential credential) {
        userService.deleteCredential(username, credential);
        return new ResponseEntity<>(this.getCredentials(username), HttpStatus.OK);
    }

    @PostMapping(value = "/updateCredential/{username}", consumes = "application/json")
    public ResponseEntity<Credential[]> updateCredential(@PathVariable("username") String username, @RequestBody CredentialHolder credentialHolder) {
        userService.updateCredential(username, credentialHolder.getOldCredential(), credentialHolder.getNewCredential());
        return new ResponseEntity<>(this.getCredentials(username), HttpStatus.OK);
    }


    private Credential[] getCredentials(String username) {
        User user = userService.getUser(username);
        List<Credential> credentialList = user.getCredentials();
        Credential[] credentials = new Credential[credentialList.size()];
        credentialList.toArray(credentials);
        return credentials;
    }

    //
    private static class CredentialHolder {
        private Credential oldCredential;
        private Credential newCredential;

        public CredentialHolder(Credential oldCredential, Credential newCredential) {
            this.oldCredential = oldCredential;
            this.newCredential = newCredential;
        }

        public Credential getOldCredential() {
            return oldCredential;
        }

        public Credential getNewCredential() {
            return newCredential;
        }
    }


    private static class LoginHolder {
        private String username;
        private String pass;

        public LoginHolder(String username, String pass) {
            this.username = username;
            this.pass = pass;
        }

        public String getUsername() {
            return username;
        }

        public String getPass() {
            return pass;
        }
    }
}
